docs: rewrite Kubernetes page (#3928)
docker-mailserver/docker-mailserver: docs/content/config/advanced/kubernetes.md

      FAIL2BAN_BLOCKTYPE: drop
      POSTMASTER_ADDRESS: postmaster@example.com
      UPDATE_CHECK_INTERVAL: 10d
      POSTFIX_INET_PROTOCOLS: ipv4
      ENABLE_CLAMAV: '1'
      ENABLE_POSTGREY: '0'
      ENABLE_FAIL2BAN: '1'
      AMAVIS_LOGLEVEL: '-1'
      SPOOF_PROTECTION: '1'
      MOVE_SPAM_TO_JUNK: '1'
      ENABLE_UPDATE_CHECK: '1'
      ENABLE_SPAMASSASSIN: '1'
      SUPERVISOR_LOGLEVEL: warn
      SPAMASSASSIN_SPAM_TO_INBOX: '1'

      # here, we provide an example for the SSL configuration
      SSL_TYPE: manual
      SSL_CERT_PATH: /secrets/ssl/rsa/tls.crt
      SSL_KEY_PATH: /secrets/ssl/rsa/tls.key
    ```

    You can also make use of user-provided configuration files (_e.g. `user-patches.sh`, `postfix-accounts.cf` and more_), to customize DMS to your needs. Here is a minimal example that supplies a `postfix-accounts.cf` file inline with two users:

    ```yaml
    ---
    apiVersion: v1
    kind: ConfigMap

    metadata:
      name: mailserver.files

    data:
      postfix-accounts.cf: |
        test@example.com|{SHA512-CRYPT}$6$someHashValueHere
        other@example.com|{SHA512-CRYPT}$6$someOtherHashValueHere
    ```

    !!! attention "Static Configuration"

>         With the inline `postfix-accounts.cf` file configured above, the content is fixed: you cannot change the configuration or persists modifications, i.e. adding or removing accounts is not possible. You need to use a `PersistentVolumeClaim` in case `postfix-accounts.cf` cannot be static.

Author: I think my version is accurate: the file is read-only, and even if you appear to have write permissions, you will not be able to actually write to the file. Mounting a `ConfigMap` or `Secret` is _always_ read-only, no modifications whatsoever.

=== "Sensitive Data"

    !!! attention "Sensitive Data"

        For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMap`s and use them the same way.

    The [TLS docs page][docs-tls] provides guidance when it comes to certificates and transport layer security. Always provide sensitive information via `Secrets`.

=== "`Deployment`"

    The `Deployment` config is the most complex component.

    - It instructs Kubernetes how to run the DMS container and how to apply your `ConfigMap`s, persisted storage, etc.
    - Additional options can be set to enforce runtime security.

    ```yaml
    ---
    apiVersion: apps/v1
    kind: Deployment

    metadata:
      name: mailserver

      annotations:
        ignore-check.kube-linter.io/run-as-non-root: >-
          'mailserver' needs to run as root
        ignore-check.kube-linter.io/privileged-ports: >-
          'mailserver' needs privileged ports
        ignore-check.kube-linter.io/no-read-only-root-fs: >-
          There are too many files written to make the root FS read-only

    spec:
      replicas: 1
      selector:
        matchLabels:

>           app: mailserver

Reviewer: The docs no longer refer to the single instance decision, [is this worth addressing](https://github.com/docker-mailserver/docker-mailserver/pull/3928#discussion_r1518574291)?

I'm fine if you don't see the point, just double checking. Can always revise in future if user feedback warrants it.

Author: I think this is not worth addressing TBH. Most users will grasp the reasons immediately, and those that need more replicas are (should be) well aware of the implications.

        app: mailserver
      annotations:
        metallb.universe.tf/address-pool: mailserver

    # ...

    ---
    apiVersion: metallb.io/v1beta1
    kind: IPAddressPool

    metadata:
      name: mail
      namespace: metallb-system

    spec:
      addresses: [ <YOUR PUBLIC, DEDICATED IP IN CIDR NOTATION> ]
      autoAssign: true

    ---
    apiVersion: metallb.io/v1beta1
    kind: L2Advertisement

    metadata:
      name: mail
      namespace: metallb-system

    spec:
      ipAddressPools: [ mailserver ]
    ```

    **Advantages / Disadvantages**

    - :+1: simple
    - :-1: requires a dedicated, publicly routable IP address
    - :-1: limited to the node with the dedicated IP address
    - :point_right: requires the setup of a load balancer

=== "External-IP Service"

    **General**

    !!! info

        This approach only works when:

        1. You can dedicate a **publicly routable IP** address to the DMS configured `Service`.
        2. The publicly routable IP is required to be dedicated to allow your mail server to have matching `A` and `PTR` records (_which other mail servers will use to verify trust when they receive mail sent from your DMS instance_).

>     In this setup, you set up the DMS configured `Service` manually with an "[external IP][Kubernetes-network-external-ip]", providing the dedicated, publicly routable IP address yourself.

Author: Not sure what to do here exactly (the suggestions seems to already have been applied? I cannot find the texts though, weird...), but apply what you deem fitting :)

>     In this setup, you set up the DMS configured `Service` manually with an "[external IP][Kubernetes-network-external-ip]", providing the dedicated, publicly routable IP address yourself.

Reviewer: > Not sure what to do here exactly (the suggestions seems to already have been applied?

I think I forgot this one sorry. I referred to a line you had since removed, and this change required me to edit the file content directly rather than a suggestion. I was out of executive functioning points by the time I pushed the larger changes, so I still need to add this suggestion, I'll take of it 👍

    This approach is very similar to the approach that uses a load balancer and a public IP address.

    **Example**

    ```yaml
    ---
    apiVersion: v1
    kind: Service

    metadata:
      name: mailserver
      labels:
        app: mailserver

    spec:
      selector:
        app: mailserver
      ports:
        - name: smtp
          port: 25
          targetPort: smtp
        # ...

      externalIPs:
        - 10.20.30.40
    ```

    **Advantages / Disadvantages**

    - :+1: simple
    - :-1: requires a dedicated, publicly routable IP address
    - :-1: limited to the node with the dedicated IP address
    - :point_right: requires manually setting the IP

=== "Host network"

    **General**

>     One way to also preserve the real client IP is to use `hostPort` and `hostNetwork: true`. This approach is similar to host network in Docker.

Author: > > One way to also preserve the real client IP
> 
> Are all examples demonstrated here preserving the real client IP? (_**EDIT:** [Your comment here](https://github.com/docker-mailserver/docker-mailserver/pull/3928#discussion_r1518586496) states you are not sure, is it difficult to verify?_)

You got a point; all the examples we provide preserve the client IP 👍🏼

> I wouldn't have thought that to be the case with the LB (_if it's the entrypoint for traffic and not just assigning IP addresses for services to bind to directly?_) Yet your pro/con lists don't seem to suggest the original client IP is lost (_as would happen with Traefik outside of k8s, which load balances to one or more container instances of a service_).

The LB is not an entry point for traffic at all, it really is just responsible for assigning special IP addresses. Hence, the LB approach also preserves the clients' IP.

Outside K8s, Traefik could also speak PROXY protocol, so there is at least to my knowledge no inherent limitation.

> If all examples preserve the client IP now, the start of this sentence is not meaningful, we can drop it.

👍🏼

        end
    ```

    !!! tip "For more information on the PROXY protocol, refer to [our dedicated docs page][docs-mailserver-behind-proxy] on the feature."

    **Advantages / Disadvantages**

    - :+1: preserves the origin IP address of clients (_which is crucial for DNS related checks_);
    - :+1: aligns with a best practice for Kubernetes by using a dedicated ingress to route external traffic to the k8s cluster (_which additionally benefits from the flexibility of routing rules_); and
    - :+1: avoids the restraint of a single [node][Kubernetes-nodes] (_as a workaround to preserve the original client IP_).
    - :point_right: requires a proxy (if you do not already have one, e.g. as an ingress, you will need to set up one)
    - :-1: added complexity
        - on the manifest side: changing the DMS configured `Service`
        - on DMS' side: changing the Postfix and Dovecot configuration
    - :-1: if you want to have cluster-internal traffic remain cluster-internal, you will need to "duplicate" the ports for Postfix and Dovecot to have ports that are PROXY-protocol enabled and ports that remain "normal"

    **Examples**

>     A complete configuration, with duplicated ports, can be found down below in the "Traefik" section. The Postfx and Dovecot configuration is identical for other proxies (like NGINX).

Author: This seems to have already happened? I will resolve this comment because the admonition is already there.

        kind: IngressRouteTCP

        metadata:
          name: imaps

        spec:
          entryPoints: [ imaps ]
          routes:
            - match: HostSNI(`*`)
              services:
                - name: mailserver
                  namespace: mail
                  port: imaps-proxy
                  proxyProtocol:
                    version: 2
        ```

        !!! info

            The `IngressRouteTCP` example configs above reference ports with a `*-proxy` suffix.

            - These port variants will be defined in the `Deployment` configuration, and are scoped to the `mailserver` service (via `spec.routes.services.name`).
            - The suffix is used to distinguish that these ports are only compatible with connections using the PROXY protocol, which is what your ingress controller should be managing for you by adding the correct PROXY protocol headers to TCP connections it routes to DMS.

        !!! warning "Connections to DMS within the internal cluster will be rejected"

            The services for these ports can only enable PROXY protocol support by mandating the protocol on all connections for these ports.

>             This can be problematic when you also need to support internal cluster traffic directly to DMS (_instead of routing indirectly through the ingress controller_).

Author: This seems to have been applied already too - hence, I will resolve this comment.

            ```yaml
            - name: smtp-proxy
              containerPort: 10025
              protocol: TCP
            - name: subs-proxy
              containerPort: 10465
              protocol: TCP
            - name: sub-proxy
              containerPort: 10587
              protocol: TCP
            - name: imaps-proxy
              containerPort: 10993
              protocol: TCP
            ```

[github-web::docker-mailserver-helm]: https://github.com/docker-mailserver/docker-mailserver-helm
[metallb-web]: https://metallb.universe.tf/

[kustomize]: https://kustomize.io/
[cert-manager]: https://cert-manager.io/docs/
[docs-tls]: ../security/ssl.md
[docs-dovecot]: ./override-defaults/dovecot.md
[docs-postfix]: ./override-defaults/postfix.md
[docs-mailserver-behind-proxy]: ../../examples/tutorials/mailserver-behind-proxy.md
[dockerhub-haproxy]: https://hub.docker.com/_/haproxy
[Kubernetes-nginx]: https://kubernetes.github.io/ingress-nginx
[Kubernetes-nginx-expose]: https://kubernetes.github.io/ingress-nginx/user-guide/exposing-tcp-udp-services
[Kubernetes-network-service]: https://kubernetes.io/docs/concepts/services-networking/service

> [Kubernetes-nodes]: https://kubernetes.io/docs/concepts/architecture/nodes[Kubernetes-network-service]: https://kubernetes.io/docs/concepts/services-networking/service

Reviewer: ```suggestion
```

      ENABLE_CLAMAV: '1'
      ENABLE_POSTGREY: '0'
      ENABLE_FAIL2BAN: '1'
      AMAVIS_LOGLEVEL: '-1'
      SPOOF_PROTECTION: '1'
      MOVE_SPAM_TO_JUNK: '1'
      ENABLE_UPDATE_CHECK: '1'
      ENABLE_SPAMASSASSIN: '1'
      SUPERVISOR_LOGLEVEL: warn
      SPAMASSASSIN_SPAM_TO_INBOX: '1'

      # here, we provide an example for the SSL configuration
      SSL_TYPE: manual
      SSL_CERT_PATH: /secrets/ssl/rsa/tls.crt
      SSL_KEY_PATH: /secrets/ssl/rsa/tls.key
    ```

    We can also make use of user-provided configuration files, e.g. `user-patches.sh`, `postfix-accounts.cf` and more, to adjust DMS to our likings. We encourage you to have a look at [Kustomize][kustomize] for creating `ConfigMap`s from multiple files, but for now, we will provide a simple, hand-written example. This example is absolutely minimal and only goes to show what can be done.

    ```yaml
    ---
    apiVersion: v1
    kind: ConfigMap

    metadata:
      name: mailserver.files

    data:
      postfix-accounts.cf: |
        test@example.com|{SHA512-CRYPT}$6$someHashValueHere
        other@example.com|{SHA512-CRYPT}$6$someOtherHashValueHere
    ```

    !!! attention "Static Configuration"

>         With the configuration shown above, you can **not** dynamically add accounts as the configuration file mounted into the mail server can not be written to.

Author: I think the wording here is not very accurate: "With the \[...\] file configured above, as the content is fixed you cannot persist modifications (_adding/removing accounts_)." That sounds a bit weird to me. I will try to re-phrase it.

        Use persistent volumes for production deployments.

=== "`PersistentVolumeClaim`"

    Thereafter, we need persistence for our data. Make sure you have a storage provisioner and that you choose the correct `storageClassName`.

    ```yaml
    ---
    apiVersion: v1
    kind: PersistentVolumeClaim

    metadata:
      name: data

    spec:
      storageClassName: local-path
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 25Gi
    ```

=== "`Service`"

>     The `Service`'s configuration determines whether the original IP from the sender will be kept. [More about this further down below](#exposing-your-mail-server-to-the-outside-world). The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most Kubernetes users will only want to have one instance.

Author: I dropped the first sentence because it's stating the obvious. In the K8s docs page, it's the equivalent of saying Docker is an application for running containers 😆 I found it to be an improvement because I am not a fan of very short paragraphs. I also cannot understand why we suddenly use bullet points for normal sentences - why do it here now? 

I will take a middle-ground now between your suggestion and my approach.

>     The `Service`'s configuration determines whether the original IP from the sender will be kept. [More about this further down below](#exposing-your-mail-server-to-the-outside-world). The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most Kubernetes users will only want to have one instance.

Author: > With my lack of experience, the `name` and `targetPort` fields having the same values is a tad unclear to me from this snippet alone 😅

`targetPort` references the port that is targeted when looking at the `Deployment`. `name` is the name of port when looking at the `Service`, which can be references from an `IngressRoute` for example.

>     The `Service`'s configuration determines whether the original IP from the sender will be kept. [More about this further down below](#exposing-your-mail-server-to-the-outside-world). The configuration you're seeing does keep the original IP, but you will not be able to scale this way. We have chosen to go this route in this case because we think most Kubernetes users will only want to have one instance.

Author: > The bullet points above are focused on the IP concern and something about how this is preserved from the config but prevents scaling beyond a single instance which from the snippet alone is not obvious why that is not scalable 🤷‍♂️ This IP concern seems more suited to be in an info admonition instead, beneath the config snippet?

Good point; I will adjust the content.

        - name: submission
          port: 587
          targetPort: submission
          protocol: TCP
        # imaps (implicit TLS)
        - name: imaps
          port: 993
          targetPort: imaps
          protocol: TCP
    ```

=== "`Deployment`"

    Last but not least, the `Deployment` becomes the most complex component. It instructs Kubernetes how to run the DMS container and how to apply your `ConfigMaps`, persisted storage, etc. Additionally, we can set options to enforce runtime security here.

    ```yaml
    ---
    apiVersion: apps/v1
    kind: Deployment

    metadata:
      name: mailserver

      annotations:
        ignore-check.kube-linter.io/run-as-non-root: >-
          'mailserver' needs to run as root
        ignore-check.kube-linter.io/privileged-ports: >-
          'mailserver' needs privileged ports
        ignore-check.kube-linter.io/no-read-only-root-fs: >-
          There are too many files written to make The
          root FS read-only

    spec:
      replicas: 1
      selector:
        matchLabels:
          app: mailserver

      template:
        metadata:
          labels:
            app: mailserver

          annotations:

>             container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default

Author: No worries as of now, we can safely still have it here. When AppArmor is GA, we can take actions.

>             container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default

Reviewer: > When AppArmor is GA, we can take actions.

I don't think it will be GA? The last link I provided is dated, while the earlier ones are quite recent.

https://github.com/kubernetes/kubernetes/pull/123435

> - _converts the AppArmor annotations to SecurityContext fields._
> - _migrate Seccomp annotations to fields:_
> - _AppArmor profiles can now be configured through fields on the PodSecurityContext and container SecurityContext._
> - _The beta AppArmor annotations are deprecated._

---

If you feel it's better to leave here in deprecated form, instead of removing or adapting to the v1.30 `SecurityContext` fields no worries then 👍

>             container.apparmor.security.beta.kubernetes.io/mailserver: runtime/default

Author: Absolutely not an issue IMO; deprecations usually provide users with a lot of time, and v1.30 is not even out at this moment.

          containers:
            - name: mailserver
              image: ghcr.io/docker-mailserver/docker-mailserver:latest
              imagePullPolicy: IfNotPresent

              securityContext:
                # Required to support SGID via `postdrop` executable
                # in `/var/mail-state` for Postfix (maildrop + public dirs):
                # https://github.com/docker-mailserver/docker-mailserver/pull/3625
                allowPrivilegeEscalation: true
                readOnlyRootFilesystem: false
                runAsUser: 0
                runAsGroup: 0
                runAsNonRoot: false
                privileged: false
                capabilities:
                  add:
                    # file permission capabilities
                    - CHOWN
                    - FOWNER
                    - MKNOD
                    - SETGID
                    - SETUID
                    - DAC_OVERRIDE
                    # network capabilities
                    - NET_ADMIN  # needed for F2B
                    - NET_RAW    # needed for F2B
                    - NET_BIND_SERVICE
                    # miscellaneous  capabilities
                    - SYS_CHROOT

>                     - KILL

Author: Leaving this unresolved for future PRs on this docs page.

              #   case you're low on resources and Kubernetes refuses to
              #   schedule new pods.
              resources:
                limits:
                  memory: 4Gi
                  cpu: 1500m
                requests:
                  memory: 2Gi
                  cpu: 600m

              volumeMounts:
                - name: files
                  subPath: postfix-accounts.cf
                  mountPath: /tmp/docker-mailserver/postfix-accounts.cf
                  readOnly: true

                # PVCs
                - name: data
                  mountPath: /var/mail
                  subPath: data
                  readOnly: false
                - name: data
                  mountPath: /var/mail-state
                  subPath: state
                  readOnly: false
                - name: data
                  mountPath: /var/log/mail
                  subPath: log
                  readOnly: false

                # certificates
                - name: certificates-rsa
                  mountPath: /secrets/ssl/rsa/
                  readOnly: true

                # other
                - name: tmp-files
                  mountPath: /tmp

>                   readOnly: false

Author: It actually is in the `volumes:` section (ll. 312-313); it's for temporary files in `/tmp`. Mounting is not special here - it's not a problem to mount a disc in `/mnt` when you have a mount for `/` too, right? it works the same here.

>                   readOnly: false

Author: Technically, it's not required because `readOnlyRootFilesystem` is `false` anyways. I will tackle this.

              ports:
                - name: smtp
                  containerPort: 25
                  protocol: TCP
                - name: submissions
                  containerPort: 465
                  protocol: TCP
                - name: submission
                  containerPort: 587
                - name: imaps
                  containerPort: 993

>                   protocol: TCP

Author: > I assume this has some relevance to the load balancer `Service` config from earlier?

Actually, this is probably something we need to clarify. When we use PROXY protocol, the service's type should be `ClusterIP`, not `LoadBalancer`.

>  * Is that connection via the `spec.selectors` field near the top of this snippet?

Yes; we match on `app: mailserver`.

>  * And the single instance I assume is handled via the `spec.replicas: 1`?

Yes; we only want one pod running DMS here.

>  * Is `name` here related to the `Service` port `name` or `targetPort` fields?

The name that we give here is referred to in the `Service` via `targetPort`. We can provide another name in the `Service`, though, via the `name:` field, if we wanted to.

              envFrom:
                - configMapRef:
                    name: mailserver.environment

          restartPolicy: Always

          volumes:
            # configuration files
            - name: files
              configMap:
                name: mailserver.files

            # PVCs
            - name: data
              persistentVolumeClaim:
                claimName: data

            # certificates
            - name: certificates-rsa
              secret:
                secretName: mail-tls-certificate-rsa
                items:
                  - key: tls.key
                    path: tls.key
                  - key: tls.crt
                    path: tls.crt

            # other
            - name: tmp-files
              emptyDir: {}
    ```

> === "Certificates"

Author: Probably a good idea!

    The [TLS docs page][docs-tls] provides guidance when it comes to certificates and transport layer security. Always provide sensitive information vai `Secrets`.

## Exposing your Mail Server to the Outside World

The more difficult part with Kubernetes is to expose a deployed DMS instance to the outside world. Kubernetes provides multiple ways for doing that; each has its upsides and downsides. The major problem with exposing DMS to the outside world in Kubernetes is to [preserve the real client IP][Kubernetes-service-source-ip]. The real client IP is required by DMS for performing IP-based DNS and spam checks.

=== "Load-Balancer + Public IP"

    This approach only works when you have a **dedicated** IP address that you can give to the responsible `Service`, e.g., with a load balancer like [MetalLB][metallb-web]. Such an IP has to be public and therefore routable. The IP is required to be dedicated to allow your mail server to have matching `A` and `PTR` records (that other mail server can checken when you send them e-mails).

>     The upside is that the manifests files and the configuration do not become more complex; the downside is that you require a dedicated IPv4 address and you are stuck to the node that has this IP address bound.

Author: 👍🏼

    For storing OpenDKIM keys, TLS certificates or any sort of sensitive data, you should be using `Secret`s. You can mount secrets like `ConfigMap`s and use them the same way.
=== "Load-Balancer + Public IP"

The [TLS docs page][docs-tls] provides guidance when it comes to certificates and transport layer security. Always provide sensitive information vai `Secrets`.
    This approach only works when you have a **dedicated** IP address that you can give to the responsible `Service`, e.g., with a load balancer like [MetalLB][metallb-web]. Such an IP has to be public and therefore routable. The IP is required to be dedicated to allow your mail server to have matching `A` and `PTR` records (that other mail server can checken when you send them e-mails).

## Exposing your Mail Server to the Outside World
    The upside is that the manifests files and the configuration do not become more complex; the downside is that you require a dedicated IPv4 address and you are stuck to the node that has this IP address bound.

> The more difficult part with Kubernetes is to expose a deployed DMS to the outside world. Kubernetes provides multiple ways for doing that; each has downsides and complexity. The major problem with exposing DMS to outside world in Kubernetes is to [preserve the real client IP][Kubernetes-service-source-ip]. The real client IP is required by DMS for performing IP-based SPF checks and spam checks. If you do not require SPF checks for incoming mails, you may disable them in your [Postfix configuration][docs-postfix] by dropping the line that states: `check_policy_service unix:private/policyd-spf`.

Author: I have added this information later: "The real client IP is required by DMS for performing IP-based DNS and spam checks." in the "Exposing your Mail Server to the Outside World" section. I think it's better to phrase it more generally.

=== "External-IP Service"

> The easiest approach was covered above, using `#!yaml externalTrafficPolicy: Local`, which disables the service proxy, but makes the service local as well (which does not scale). This approach only works when you are given the correct (that is, a public and routable) IP address by a load balancer (like MetalLB). In this sense, the approach above is similar to the next example below. We want to provide you with a few alternatives too. **But** we also want to communicate the idea of another simple method: you could use a load-balancer without an external IP and DNAT the network traffic to the mail server. After all, this does not interfere with SPF checks because it keeps the origin IP address. If no dedicated external IP address is available, you could try the latter approach, if one is available, use the former.

Author: `externalTrafficPolicy` was actually outdated because it was removed entirely with an earlier PR.

> But now some of that context is missing with changes proposed by this PR?

I removed what I thought to be superfluous; the information in this section is accurate now, but shorter. I am not worried about this.

> Presumably this is part of the "Load-Balancer + Public IP" tab approach, as the paragraph then continues to refer to the same approach but stating the public routable IP requirement for a load balancer and using MetalLB as an example. It refers to "approach above" and the "next example" (now "External-IP Service" tab) being similar, although I'm not getting much value as the reader from that.

I have rephrased one paragraph a bit now; but where is the "next example" line? I could not find it...

> The easiest approach was covered above, using `#!yaml externalTrafficPolicy: Local`, which disables the service proxy, but makes the service local as well (which does not scale). This approach only works when you are given the correct (that is, a public and routable) IP address by a load balancer (like MetalLB). In this sense, the approach above is similar to the next example below. We want to provide you with a few alternatives too. **But** we also want to communicate the idea of another simple method: you could use a load-balancer without an external IP and DNAT the network traffic to the mail server. After all, this does not interfere with SPF checks because it keeps the origin IP address. If no dedicated external IP address is available, you could try the latter approach, if one is available, use the former.

Reviewer: > where is the "next example" line? I could not find it...

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/b569c280-64ab-4fe7-9609-4defb39f736d)

> The easiest approach was covered above, using `#!yaml externalTrafficPolicy: Local`, which disables the service proxy, but makes the service local as well (which does not scale). This approach only works when you are given the correct (that is, a public and routable) IP address by a load balancer (like MetalLB). In this sense, the approach above is similar to the next example below. We want to provide you with a few alternatives too. **But** we also want to communicate the idea of another simple method: you could use a load-balancer without an external IP and DNAT the network traffic to the mail server. After all, this does not interfere with SPF checks because it keeps the origin IP address. If no dedicated external IP address is available, you could try the latter approach, if one is available, use the former.

Author: Ah, it was in the already removed diff; thanks for the clarification!

>     Another simple way is to expose DMS as a `Service` with [external IPs][Kubernetes-network-external-ip]. This approach is very similar to the former approach. Here, an external IP is given to the service directly by you. With the approach above, you tell your load-balancer to do this.

Author: We could probably add a config snippet, but I'd rather like to do that later TBH because I don't want to introduce more unwarranted noise into an already big PR.

I have opted for "previous approach" and added the name in parenthesis.

>     This approach has the same upsides and downside as the former approach.

Author: I have now applied "previous approach" more consistently (and removed "former").

About the preservation of the IP: I don't know either; IIRC I did not write this, and I thought it to be rather confusing. The whole "Proxy Pod" section is unnecessary IMO.

>     This approach has the same upsides and downside as the former approach.

Author: See [4d68958](https://github.com/docker-mailserver/docker-mailserver/pull/3928/commits/4d68958c3f8e429b02fe3988c192d745ae733997); this should be solved now.

### External IPs Service
    ```yaml
    ---
    apiVersion: v1
    kind: Service

The simplest way is to expose DMS as a [Service][Kubernetes-network-service] with [external IPs][Kubernetes-network-external-ip]. This is very similar to the approach taken above. Here, an external IP is given to the service directly by you. With the approach above, you tell your load-balancer to do this.
    metadata:
      name: mailserver
      labels:
        app: mailserver

```yaml
---
apiVersion: v1
kind: Service

metadata:
  name: mailserver
  labels:
    app: mailserver

spec:
  selector:
    app: mailserver
  ports:
    - name: smtp
      port: 25
      targetPort: smtp
    # ...
    spec:
      selector:
        app: mailserver
      ports:
        - name: smtp
          port: 25
          targetPort: smtp
        # ...

  externalIPs:
    - 80.11.12.10
```
      externalIPs:
        - 80.11.12.10
    ```

This approach
    This approach has the same upsides and downside as the former approach.

- does not preserve the real client IP, so SPF check of incoming mail will fail.
- requires you to specify the exposed IPs explicitly.
=== "Host network"

> ### Proxy port to Service

Author: Yes; I have never heard of this being used in production, and most certainly, people will get confused when another tab is about PROXY protocol. Better to keep to obvious and straight-forward approaches.

>     One way to also preserve the real client IP is to use `hostPort` and `hostNetwork: true`. With this approach, you bind DMS to a specific node, but also benefit from reduced complexity. Moreover, it is not possible to access DMS via other cluster nodes, only via the node that DMS was deployed on. Additionally, every Port within the container is exposed on the host side.

Author: I have now decided to give this whole section another overhaul again; this should solve the mentioned issues altogether.

>     One way to also preserve the real client IP is to use `hostPort` and `hostNetwork: true`. With this approach, you bind DMS to a specific node, but also benefit from reduced complexity. Moreover, it is not possible to access DMS via other cluster nodes, only via the node that DMS was deployed on. Additionally, every Port within the container is exposed on the host side.

Author: ref: [4d68958](https://github.com/docker-mailserver/docker-mailserver/pull/3928/commits/4d68958c3f8e429b02fe3988c192d745ae733997)

>     kind: Deployment

Author: Use `apps/v1` 😆 I will correct this.

>     Using the PROXY protocol comes at the cost of added complexity, both on the manifest side as well as on the configuration side of DMS itself. Additionally, if you want to have cluster-internal traffic remain cluster-internal, you will need to "duplicate" the ports for Postfix and Dovecot to have ports that are PROXY-protocol enabled and ports that remain "normal". Such a configuration, with duplicated ports, can be found down below in the "Traefik" section.

Author: I have already revised the whole section, hence this looks all different and better already.

            - name: smtp-auth
              containerPort: 587
              hostPort: 587
            - name: imap-secure
              containerPort: 993
              hostPort: 993
        #  ...
```
    The PROXY protocol "wraps" incoming flows and marks them as "wrapped". This allows DMS to "unwrap" the packages and work with the original IP addresses.

With this approach,
    Additional documentation, independent of Kubernetes, can be found [here][docs-mailserver-behind-proxy].

- it is not possible to access DMS via other cluster Nodes, only via the Node DMS was deployed at.
- every Port within the Container is exposed on the Host side.
    **Drawbacks**

### Proxy Port to Service via PROXY Protocol
    Using the PROXY protocol comes at the cost of added complexity, both on the manifest side as well as on the configuration side of DMS itself. Additionally, if you want to have cluster-internal traffic remain cluster-internal, you will need to "duplicate" the ports for Postfix and Dovecot to have ports that are PROXY-protocol enabled and ports that remain "normal". Such a configuration, with duplicated ports, can be found down below in the "Traefik" section.

> This way is ideologically the same as [using a proxy pod](#proxy-port-to-service), but instead of a separate proxy pod, you configure your ingress to proxy TCP traffic to the DMS pod using the PROXY protocol, which preserves the real client IP.

Author: Indeed :)

    === "NGINX"

#### Configure your Ingress
        With an [NGINX ingress controller][Kubernetes-nginx], add the following to the TCP services config map (as described [here][Kubernetes-nginx-expose]):

> With an [NGINX ingress controller][Kubernetes-nginx], set `externalTrafficPolicy: Local` for its service, and add the following to the TCP services config map (as described [here][Kubernetes-nginx-expose]):

Author: It is absolutely not relevant anymore, indeed!

>         kind: IngressRouteTCP

Author: Probably a good idea. I will update my own cluster and can then come back to our docs.

>         === "Cluster-Internal E-Mails Not Required"
> 
>             Here is an exmaple configuration for [Postfix][docs-postfix], [Dovecot][docs-dovecot], and the `Deployment`:

Reviewer: ```suggestion
        === "Only accept connections with PROXY protocol"

            Here is an example configuration for [Postfix][docs-postfix], [Dovecot][docs-dovecot], and the adjustments to the `Deployment` config. The port names are adjusted here only for the additional context as described above.
```

>                   }

Reviewer: You're not configuring `imap-proxy` below, why is this here?

What is with the `127.0.0.0/8` network for `haproxy_trusted_networks`? DMS isn't speaking PROXY protocol to itself on loopback? Was this actually required for something? It's not present in the separate ports config tab or in either of the dedicated PROXY protocol docs page examples.

I'm still not a fan of an entire subnet as is encouraged here with the pod CIDR, but I understand that's a k8s concern. You just have to trust that there is no networking surprises that permit traffic to slip through and inject their own PROXY protocol header.

Author: > You're not configuring `imap-proxy` below, why is this here?

Can you elaborate? I am not sure if I understand.

> What is with the `127.0.0.0/8` network for `haproxy_trusted_networks`? DMS isn't speaking PROXY protocol to itself on loopback? Was this actually required for something? It's not present in the separate ports config tab or in either of the dedicated PROXY protocol docs page examples.

We can remove the 127.0.0.1/8, it serves no purpose.

> I'm still not a fan of an entire subnet as is encouraged here with the pod CIDR, but I understand that's a k8s concern. You just have to trust that there is no networking surprises that permit traffic to slip through and inject their own PROXY protocol header.

It's actually required because the ingress pod IP can change at any point of the pod is destroyed (for whatever reason). Using the Pod-CIDR here is absolutely correct.

>         === "Cluster-Internal E-Mails Required"

Reviewer: ```suggestion
        === "Separate PROXY protocol ports for ingress"
```

>             ```

Author: I think it's fitting as long as the sentence covering the snippet says "extending".

>             ```

Reviewer: Extending the `Deployment` config makes sense in that this snippet would append/merge to that earlier example.
- But that already has the regular ports configured, whereas I think the intention here is to replace them?
- Or were you referring to extending the `Deployment` config of the sibling tab? It's a little vague which one is referred to here.

Or is this section of config relevant to other containers within the cluster, not only for the ingress?

> [docs-mailserver-behind-proxy]: ../../../examples/tutorials/mailserver-behind-proxy

Reviewer: Excess `../` + `.md` extension for a page is usually required AFAIK?:

```suggestion
[docs-mailserver-behind-proxy]: ../../examples/tutorials/mailserver-behind-proxy.md
```

              protocol: TCP
            - name: subs-proxy
              containerPort: 10465
              protocol: TCP
            - name: sub-proxy
              containerPort: 10587
              protocol: TCP
            - name: imaps-proxy
              containerPort: 10993
              protocol: TCP
            ```

- it is not possible to access DMS via cluster-DNS, as the PROXY protocol is required for incoming connections.
[github-web::docker-mailserver-helm]: https://github.com/docker-mailserver/docker-mailserver-helm
[metallb-web]: https://metallb.universe.tf/

[Helm chart]: https://github.com/docker-mailserver/docker-mailserver-helm
[kustomize]: https://kustomize.io/
[cert-manager]: https://cert-manager.io/docs/
[docs-tls]: ../security/ssl.md
[docs-dovecot]: ./override-defaults/dovecot.md
[docs-postfix]: ./override-defaults/postfix.md
[docs-mailserver-behind-proxy]: ../../../examples/tutorials/mailserver-behind-proxy
[dockerhub-haproxy]: https://hub.docker.com/_/haproxy
[Kubernetes-nginx]: https://kubernetes.github.io/ingress-nginx
[Kubernetes-nginx-expose]: https://kubernetes.github.io/ingress-nginx/user-guide/exposing-tcp-udp-services

> [Kubernetes-network-service]: https://kubernetes.io/docs/concepts/services-networking/service

Reviewer: Bring this back, referenced in "Service" tab suggestion.

[Kubernetes-network-external-ip]: https://kubernetes.io/docs/concepts/services-networking/service/#external-ips

> [Kubernetes-nodes]: https://kubernetes.io/docs/concepts/architecture/nodes

Reviewer: Likewise restored via a suggestion.
